packages/integration_test/lib/src/vm_service_golden_client.dart

```diff
@@ -0,0 +1,226 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Examples can assume:
+// import 'dart:developer' as dev;
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:developer' as dev;
+import 'dart:io' as io;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:meta/meta.dart';
+
+/// Compares image pixels against a golden image file on the host system.
+///
+/// This comparator will send a request, using the VM service protocol, to a
+/// host script (i.e. the _driver_ script, running in a Dart VM on the host
+/// desktop OS), which will then forward the comparison to a concrete
+/// [GoldenFileComparator].
+///
+/// To use, run [useIfRunningOnDevice] in the `main()` of a test file or similar:
+///
+/// ```dart
+/// import 'package:integration_test/integration_test.dart';
+///
+/// void main() {
+///   VmServiceProxyGoldenFileComparator.useIfRunningOnDevice();
+///
+///   // Actual tests and such below.
+/// }
+/// ```
+///
+/// When either [compare] or [update] is called, the following event is sent
+/// with [dev.postEvent]:
+///
+/// ```dart
+/// dev.postEvent('compare' /* or 'update' */, <String, Object?>{
+///   'id':    1001,                 // a valid unique integer, often incrementing;
+///   'path':  'path/to/image.png',  // golden key created by matchesGoldenFile;
+///   'bytes': '...base64encoded',   // base64 encoded bytes representing the current image.
+/// }, stream: 'integration_test.VmServiceProxyGoldenFileComparator');
+/// ```
+///
+/// The comparator expects a response at the service extension
+/// `ext.integration_test.VmServiceProxyGoldenFileComparator` that is either
+/// of the following formats:
+///
+/// ```dart
+/// <String, Object?>{
+///   'error': 'Description of why the operation failed'
+/// }
+/// ```
+///
+/// or:
+///
+/// ```dart
+/// <String, Object?>{
+///   'result': true /* or possibly false, in the case of 'compare' calls */
+/// }
+/// ```
+///
+/// See also:
+///
+///   * [matchesGoldenFile], the function that invokes the comparator.
+@experimental
+final class VmServiceProxyGoldenFileComparator extends GoldenFileComparator {
+  VmServiceProxyGoldenFileComparator._() : _postEvent = dev.postEvent {
+    dev.registerExtension(_kServiceName, (_, Map<String, String> parameters) {
+      return handleEvent(parameters);
+    });
+  }
+
+  /// Creates an instance of [VmServiceProxyGoldenFileComparator] for internal testing.
+  ///
+  /// @nodoc
+  @visibleForTesting
+  VmServiceProxyGoldenFileComparator.forTesting(this._postEvent);
+
+  static bool get _isRunningOnHost {
+    if (kIsWeb) {
+      return false;
+    }
+    return !io.Platform.isAndroid && !io.Platform.isIOS;
```

Comment:
Are these the only platforms that separate device from host?

Reply:
Currently, yes.